mayastor/src/bdev/nexus/nexus_io.rs

```diff
@@ -516,54 +596,88 @@ impl NexusBio {
-                        nexus.reconfigure(DrEvent::ChildFault).await;
-                        // TODO: an error can occur here if a separate task,
-                        // e.g. grpc request is also deleting the child.
-                        if let Err(err) = child.destroy().await {
-                            error!(
-                                "{}: destroying child {} failed {}",
-                                nexus, child, err
+                match child_state {
+                    None => {
+                        debug!(
+                            "{} does not belong (anymore) to nexus {}",
+                            device, nexus
+                        );
+                    }
+                    Some(current_state) => {
+                        if current_state == ChildState::Open {
+                            warn!(
+                                "core {} thread {:?}, faulting child {}",
+                                Cores::current(),
+                                Mthread::current(),
+                                device,
                             );
-                        }
 
-                        nexus.resume().await.unwrap();
-                        if nexus.status() == NexusStatus::Faulted {
-                            error!(":{} has no children left... ", nexus);
+                            // Pausing a nexus acts like entering a critical
+                            // section,
+                            // allowing only one retire request to run at a
+                            // time, which prevents
+                            // inconsistency in reading/updating nexus
+                            // configuration.
+                            nexus.pause().await.unwrap();
+                            nexus.set_failfast().await.unwrap();
+                            nexus.reconfigure(DrEvent::ChildFault).await;
+
+                            // Lookup child once more and finally remove it.
+                            match nexus.child_lookup(&device) {
+                                Some(child) => {
+                                    // TODO: an error can occur here if a
+                                    // separate task,
+                                    // e.g. grpc request is also deleting the
+                                    // child.
+                                    if let Err(err) = child.destroy().await {
+                                        error!(
+                                            "{}: destroying child {} failed {}",
+                                            nexus, child, err
+                                        );
+                                    }
+                                }
+                                None => {
+                                    warn!(
+                                        "{} does not belong (anymore) to nexus {}, skipping child removal",
```

Comment:
Slightly clearer as "... no longer belongs to nexus ...".

Reply:
Fixed.